api/client/lib.py

```diff
@@ -236,26 +236,26 @@ def get_data(url, headers, params=None, logger=None):
         log_record['elapsed_time_in_ms'] = 1000 * elapsed_time
         log_record['retry_count'] = retry_count
         log_record['status_code'] = response.status_code
-        if response.status_code == 200:
+        if response.status_code == 200:  # Success
             logger.debug('OK', extra=log_record)
             return response
-        if response.status_code == 204:
-            logger.warning('No Content', extra=log_record)
+        if response.status_code in [204, 206]:  # Success with a caveat - warning
+            log_msg = {204: 'No Content', 206: 'Partial Content'}[response.status_code]
+            logger.warning(log_msg, extra=log_record)
             return response
         retry_count += 1
         log_record['tag'] = 'failed_gro_api_request'
         if retry_count < cfg.MAX_RETRIES:
             logger.warning(response.text, extra=log_record)
-        if response.status_code == 429:
-            time.sleep(2 ** retry_count)  # Exponential backoff before retrying
-        elif response.status_code == 301:
+        if response.status_code in [400, 401, 402, 404]:
+            break  # Do not retry
+        if response.status_code == 301:
             new_params = redirect(params, response.json()['data'][0])
             logger.warning('Redirecting {} to {}'.format(params, new_params), extra=log_record)
             params = new_params
-        elif response.status_code in [400, 401, 404, 500]:
-            break
         else:
-            logger.error('{}'.format(response), extra=log_record)
+            logger.warning('{}'.format(response), extra=log_record)
+            time.sleep(2 ** retry_count)  # Exponential backoff before retrying
```

Comment:
hmm not sure we want exponential backoff in all cases, immediate retry is. why not keep 429 as a special case?

Reply:
It's a trade-off, right? Immediate retries maximize potential throughput, but adding backoffs increases the likelihood that a request will succeed eventually without the user needing to reinvent the wheel with their own retrying.

If the issue's limited to a single load-balanced node or one random dropped connection, immediate retry would be more optimal, but sometimes a process might be down momentarily and immediately responding errors - good to give it a chance to start back up instead of getting 5 rapid errors in a row and crashing.

A minor optimization: we could get the best of both worlds by doing one immediate retry and then only backing off on subsequent ones, so at least it doesn't add that unnecessary second of delay to a one-off error.

Reply:
^ added immediate retry for first failure